src/lib/auth.ts

```diff
@@ -339,50 +344,42 @@ export const authenticate = (appMode: AppMode, appConfig: AppConfigModelType, ur
-            return getFirebaseJWTWithBearerToken(basePortalUrl, "Bearer", bearerToken, classInfo.classHash)
-              .then(([rawFirebaseJWT, firebaseJWT]) => {
-                getPortalOfferings(portalJWT.user_type, portalJWT.uid, portalJWT.domain, rawPortalJWT)
-                  .then(result => {
-                    portalClassOfferings = getPortalClassOfferings(result, appConfig, urlParams);
-                    const uidAsString = `${portalJWT.uid}`;
-                    let authenticatedUser: AuthenticatedUser | undefined;
-
-                    if (portalJWT.user_type === "learner") {
-                      authenticatedUser = classInfo.students.find((student) => student.id === uidAsString);
-                    } else {
-                      authenticatedUser = classInfo.teachers.find((teacher) => teacher.id === uidAsString);
-                    }
-
-                    if (!authenticatedUser) {
-                      throw new Error("Current user not found in class roster");
-                    }
-
-                    authenticatedUser.portalJWT = portalJWT;
-                    authenticatedUser.rawPortalJWT = rawPortalJWT;
-                    authenticatedUser.firebaseJWT = firebaseJWT;
-                    authenticatedUser.rawFirebaseJWT = rawFirebaseJWT;
-                    authenticatedUser.id = uidAsString;
-                    authenticatedUser.portal = portal;
-                    if (portalClassOfferings) {
-                      authenticatedUser.portalClassOfferings = portalClassOfferings;
-                    }
-                    getProblemIdForAuthenticatedUser(rawPortalJWT, appConfig, urlParams)
-                      .then(({ unitCode: newUnitCode, problemOrdinal: newProblemOrdinal }) => {
-                        if (authenticatedUser) {
-                          Logger.log(LogEventName.INTERNAL_AUTHENTICATED, {id: authenticatedUser.id, portal});
-                          resolve({
-                            authenticatedUser,
-                            classInfo,
-                            unitCode: newUnitCode,
-                            problemId: newProblemOrdinal
-                          });
-                        }
-                      })
-                      .catch(reject);
-                  })
-                  .catch(reject);
-              })
-              .catch(reject);
+            const { user_type, uid, domain } = portalJWT;
+            const { classHash } = classInfo;
+            const uidAsString = `${portalJWT.uid}`;
+            const firebaseJWTPromise = getFirebaseJWTWithBearerToken(basePortalUrl, "Bearer", bearerToken, classHash);
+            const portalOfferingsPromise = getPortalOfferings(user_type, uid, domain, rawPortalJWT);
+            const problemIdPromise = getProblemIdForAuthenticatedUser(rawPortalJWT, appConfig, urlParams);
+            Promise
+              .all([firebaseJWTPromise, portalOfferingsPromise, problemIdPromise])
```

Comment:
At an earlier point in the development of this PR, I thought it would make sense to read the firestore user document here where we gather the rest of the user information. This code was already nested five promises deep and so rather than make it six, I determined that several of the promises in question were mutually independent and so could be run in parallel rather than sequentially. Therefore, I refactored the code to use `Promise.all()` only to then determine that we can't read the firestore user document here because firestore hasn't been initialized at this point. But the `Promise.all()` refactor seemed worth keeping since it parallelizes the authentication process.